File: app/src/main/java/com/example/myapplication/Repositories/TemplateRepository.java
package com.example.myapplication.Repositories;

import android.app.Application;
import android.os.AsyncTask;
import android.renderscript.Sampler;

import androidx.lifecycle.LiveData;

import com.example.myapplication.Daos.TemplateDao;
import com.example.myapplication.Databases.CollageDatabase;
import com.example.myapplication.Models.Template;

import java.util.List;
import java.util.concurrent.ExecutionException;

public class TemplateRepository {
    private TemplateDao templateDao;
    private LiveData<List<Template>> allTemplates;

    public TemplateRepository(Application application) {
        CollageDatabase database = CollageDatabase.getInstance(application);
        templateDao = database.templateDao();
        allTemplates = templateDao.getAllTemplates();
    }

    public void insert(Template template) {
        new InsertTemplateAsyncTask(templateDao).execute(template);
    }

    public void update(Template template) {
        new UpdateTemplateAsyncTask(templateDao).execute(template);
    }

    public void delete(Template template) {
        new DeleteTemplateAsyncTask(templateDao).execute(template);
    }

    public Template getTemplateById(int id) throws ExecutionException, InterruptedException {

        return new GetTemplateByIdAsyncTask(templateDao).execute(id).get();
    }

    public LiveData<List<Template>> getAllTemplates() {
        return allTemplates;
    }

    private static class InsertTemplateAsyncTask extends AsyncTask<Template, Void, Void> {

        private TemplateDao templateDao;

        private InsertTemplateAsyncTask(TemplateDao templateDao) {
            this.templateDao = templateDao;
        }

        @Override
        protected Void doInBackground(Template... templates) {
            templateDao.insert(templates[0]);
            return null;
        }
    }

    private static class UpdateTemplateAsyncTask extends AsyncTask<Template, Void, Void> {

        private TemplateDao templateDao;

        private UpdateTemplateAsyncTask(TemplateDao templateDao) {
            this.templateDao = templateDao;
        }

        @Override
        protected Void doInBackground(Template... templates) {
            templateDao.update(templates[0]);
            return null;
        }
    }

    private static class DeleteTemplateAsyncTask extends AsyncTask<Template, Void, Void> {

        private TemplateDao templateDao;

        private DeleteTemplateAsyncTask(TemplateDao templateDao) {
            this.templateDao = templateDao;
        }

        @Override
        protected Void doInBackground(Template... templates) {
            templateDao.delete(templates[0]);
            return null;
        }
    }

    private class GetTemplateByIdAsyncTask extends AsyncTask<Integer, Void, Template> {

        private TemplateDao templateDao;

        private GetTemplateByIdAsyncTask(TemplateDao templateDao) {
            this.templateDao = templateDao;
        }

        @Override
        protected Template doInBackground(Integer... integers) {
            return templateDao.getTemplateById(integers[0]);
        }

        @Override
        protected void onPostExecute(Template template) {
        }
    }

}

File: app/src/main/java/com/example/myapplication/repositories/TemplateRepository.kt
package com.example.myapplication.repositories

import android.app.Application
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import com.example.myapplication.daos.TemplateDao
import com.example.myapplication.databases.CollageDatabase
import com.example.myapplication.models.Template
import java.util.concurrent.ExecutionException

class TemplateRepository(application: Application?) {
    private val templateDao: TemplateDao
    val allTemplates: LiveData<List<Template>>
    fun insert(template: Template?) {
        InsertTemplateAsyncTask(templateDao).execute(template)
    }

    fun update(template: Template?) {
        UpdateTemplateAsyncTask(templateDao).execute(template)
    }

    fun delete(template: Template?) {
        DeleteTemplateAsyncTask(templateDao).execute(template)
    }

    @Throws(ExecutionException::class, InterruptedException::class)
    fun getTemplateById(id: Int): Template {
        return GetTemplateByIdAsyncTask(templateDao).execute(id).get()
    }

    private class InsertTemplateAsyncTask private constructor(private val templateDao: TemplateDao) : AsyncTask<Template?, Void?, Void?>() {
        protected override fun doInBackground(vararg templates: Template): Void? {
            templateDao.insert(templates[0])
            return null
        }

    }

    private class UpdateTemplateAsyncTask private constructor(private val templateDao: TemplateDao) : AsyncTask<Template?, Void?, Void?>() {
        protected override fun doInBackground(vararg templates: Template): Void? {
            templateDao.update(templates[0])
            return null
        }

    }

    private class DeleteTemplateAsyncTask private constructor(private val templateDao: TemplateDao) : AsyncTask<Template?, Void?, Void?>() {
        protected override fun doInBackground(vararg templates: Template): Void? {
            templateDao.delete(templates[0])
            return null
        }

    }

    private inner class GetTemplateByIdAsyncTask private constructor(private val templateDao: TemplateDao) : AsyncTask<Int?, Void?, Template>() {
        protected override fun doInBackground(vararg integers: Int): Template {
            return templateDao.getTemplateById(integers[0])
        }

        override fun onPostExecute(template: Template) {}

    }

    init {
        val database: CollageDatabase = CollageDatabase.getInstance(application)
        templateDao = database.templateDao()
        allTemplates = templateDao.allTemplates
    }
}